File: doc/src/scripts/TreeTraversal.cpp
#include <krpc.hpp>
#include <krpc/services/space_center.hpp>
#include <iostream>
#include <stack>

using namespace krpc;
using namespace krpc::services;

int main() {
  Client conn = krpc::connect("");
  SpaceCenter sc(&conn);
  auto vessel = sc.active_vessel();

  auto root = vessel.parts().root();
  std::stack<std::pair<SpaceCenter::Part,int> > stack;
  stack.push(std::pair<SpaceCenter::Part,int>(root, 0));
  while (stack.size() > 0) {
    auto part = stack.top().first;
    auto depth = stack.top().second;
    stack.pop();
    std::cout << std::string(depth, ' ') << part.title() << std::endl;
    auto children = part.children();
    for (std::vector<SpaceCenter::Part>::iterator child = children.begin(); child != children.end(); child++) {
      stack.push(std::pair<SpaceCenter::Part,int>(*child, depth+1));
    }
  }
}

File: doc/src/scripts/TreeTraversal.py
import krpc
conn = krpc.connect()
vessel = conn.space_center.active_vessel

root = vessel.parts.root
stack = [(root, 0)]
while len(stack) > 0:
    part, depth = stack.pop()
    print(' '*depth, part.title)
    for child in part.children:
        stack.append((child, depth+1))
